### ISSUE TITLE
Failed to update DistributedTransformerTest to use GPT-3 sizes.

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/Fuser#2962]
For a fair comparison with the baseline in test_transformer_engine.py. 

I tried the following briefly and hit big number mismatches. 

```patch
diff --git a/tests/cpp/test_multidevice_transformer.cpp b/tests/cpp/test_multidevice_transformer.cpp
index af1187d3..2c922418 100644
--- a/tests/cpp/test_multidevice_transformer.cpp
+++ b/tests/cpp/test_multidevice_transformer.cpp
@@ -17,7 +17,7 @@

 namespace nvfuser {

-constexpr int64_t B = 2, E = 768, H = 16, S = 128;
+constexpr int64_t B = 1, E = 12288, H = 96, S = 2048;
 // Note parameters scaled by kParamScale following weight initialization
 // recommendations:
 // https://huggingface.co/docs/transformers/en/model_doc/gpt2#transformers.GPT2Config.initializer_range
```

```
$ _bn && mpirun -np 4 bin/test_multidevice --gtest_filter=DistributedTransformerTest.Forward/__half

Note: Google Test filter = DistributedTransformerTest.Forward/__half
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DistributedTransformerTest
[ RUN      ] DistributedTransformerTest.Forward/__half
Note: Google Test filter = DistributedTransformerTest.Forward/__half
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DistributedTransformerTest
[ RUN      ] DistributedTransformerTest.Forward/__half
Note: Google Test filter = DistributedTransformerTest.Forward/__half
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DistributedTransformerTest
[ RUN      ] DistributedTransformerTest.Forward/__half
Note: Google Test filter = DistributedTransformerTest.Forward/__half
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DistributedTransformerTest
[ RUN      ] DistributedTransformerTest.Forward/__half
/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 1 mismatches with atol 0.01:
  max absolute error under rtol: 2.74754
  max relative error: 0.121809
  failing elements: 4182382, 16.6193% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 1 mismatches with atol 0.01:
  max absolute error under rtol: 2.74754
  max relative error: 0.121809
  failing elements: 4182382, 16.6193% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 1 mismatches with atol 0.01:
  max absolute error under rtol: 2.74754
  max relative error: 0.121809
  failing elements: 4182382, 16.6193% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 1 mismatches with atol 0.01:
  max absolute error under rtol: 2.74754
  max relative error: 0.121809
  failing elements: 4182382, 16.6193% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 2 mismatches with atol 0.01:
  max absolute error under rtol: 2.37444
  max relative error: 0.413194
  failing elements: 4179695, 16.6086% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 2 mismatches with atol 0.01:
  max absolute error under rtol: 2.37444
  max relative error: 0.413194
  failing elements: 4179695, 16.6086% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 2 mismatches with atol 0.01:
  max absolute error under rtol: 2.37444
  max relative error: 0.413194
  failing elements: 4179695, 16.6086% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 2 mismatches with atol 0.01:
  max absolute error under rtol: 2.37444
  max relative error: 0.413194
  failing elements: 4179695, 16.6086% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 3 mismatches with atol 0.02:
  max absolute error under rtol: 41.4099
  max relative error: 0.89948
  failing elements: 23659264, 94.0135% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 3 mismatches with atol 0.02:
  max absolute error under rtol: 41.4099
  max relative error: 0.89948
  failing elements: 23659264, 94.0135% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 3 mismatches with atol 0.02:
  max absolute error under rtol: 41.4099
  max relative error: 0.89948
  failing elements: 23659264, 94.0135% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 3 mismatches with atol 0.02:
  max absolute error under rtol: 41.4099
  max relative error: 0.89948
  failing elements: 23659264, 94.0135% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 4 mismatches with atol 0.02:
  max absolute error under rtol: 41.4098
  max relative error: 0.884953
  failing elements: 23706216, 94.2% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 4 mismatches with atol 0.02:
  max absolute error under rtol: 41.4098
  max relative error: 0.884953
  failing elements: 23706216, 94.2% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 4 mismatches with atol 0.02:
  max absolute error under rtol: 41.4098
  max relative error: 0.884953
  failing elements: 23706216, 94.2% of tensor

/opt/pytorch/nvfuser/tests/cpp/test_multidevice_transformer.cpp:107: Failure
Value of: outputs[i].allclose(expected_outputs[i], rtol, atol)
  Actual: false
Expected: true
Output 4 mismatches with atol 0.02:
  max absolute error under rtol: 41.4098
  max relative error: 0.884953
  failing elements: 23706216, 94.2% of tensor

To reproduce: NVFUSER_TEST_RANDOM_SEED=1726700667 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='DistributedTransformerTest.Forward/__half'
To reproduce: NVFUSER_TEST_RANDOM_SEED=1726700667 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='DistributedTransformerTest.Forward/__half'
To reproduce: NVFUSER_TEST_RANDOM_SEED=1726700667 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='DistributedTransformerTest.Forward/__half'
To reproduce: NVFUSER_TEST_RANDOM_SEED=1726700667 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='DistributedTransformerTest.Forward/__half'
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half (4898 ms)
[----------] 1 test from DistributedTransformerTest (4898 ms total)

[----------] Global test environment tear-down
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half (4893 ms)
[----------] 1 test from DistributedTransformerTest (4893 ms total)

[----------] Global test environment tear-down
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half (4876 ms)
[----------] 1 test from DistributedTransformerTest (4876 ms total)

[----------] Global test environment tear-down
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half (4894 ms)
[----------] 1 test from DistributedTransformerTest (4894 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (5270 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half

 1 FAILED TEST
[==========] 1 test from 1 test suite ran. (5298 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half

 1 FAILED TEST
[==========] 1 test from 1 test suite ran. (5294 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half

 1 FAILED TEST
[==========] 1 test from 1 test suite ran. (5401 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DistributedTransformerTest.Forward/__half, where GetParam() = __half

 1 FAILED TEST
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[21564,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```

**Comment by wujingyue** [comment]
cc @cowanmeg 

I'm not saying we should drop everything to fix this, but 88% relative error looks really bad and can hide real errors. 

Remember you scaled down the weights? I wonder that scale needs to be adjusted for larger sizes and whether we should scale down input activations as well.

**Comment by cowanmeg** [comment]
We probably need to scale the activations as well. The magnitude difference between the weight and activations might be causing a lot of error.

**Comment by wujingyue** [comment]
Before I forget about this, https://github.com/Lightning-AI/litgpt/blob/main/tutorials/finetune_full.md is the instruction to finetune a model. We can do that to collect real activations and weights, or even just the scales of them.

**Comment by wujingyue** [comment]
FWIW, https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/distributed/run_numerics.py#L689 is how TE distributed tests verify numerics of a TransformerLayer.

**Comment by cowanmeg** [comment]
Yes, as I've mentioned the larger issue is an accumulation of error.... they are syncing the reference and test inputs for each operation to avoid this.

**Comment by wujingyue** [comment]
> they are syncing the reference and test inputs for each operation to avoid this.

Where did they do that? https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/distributed/run_numerics.py#L643-L658 appears to be running a te.TransformerLayer on single GPU and multiple GPUs separately. I noticed it syncs [inputs](https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/distributed/run_numerics.py#L671) and [weights](https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/distributed/run_numerics.py#L660), but I didn't notice it tries to sync the intermediates. Am I missing something?

**Comment by cowanmeg** [comment]
My bad, I interpreted a comment about synchronizing inputs and params.

Regardless, please note that these tests deliberately use small sizes for any reduction axis (hidden size=64, ffn_hidden_size=32) because as the comment mentions, larger sizes naturally lead to more error. https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/distributed/run_numerics.py#L639C5-L641C77. 

Our current tests use a hidden_size=768 and ffn_hidden size of 4*678, and this issue wants to increase the hidden_size to 12288 and ffn_hidden_size to 4* 12288. 

We can implement the reference in double and aggressively cast back and forth. In fact previous versions of the tests did perform the reference computation in double. However, if you want the test to use large input and parameter sizes, I don't think this method is enough to keep the tolerances small and have passing tests.

**Comment by wujingyue** [comment]
Note for myself: revisit after https://github.com/NVIDIA/Fuser/issues/3194
